tvx/drivers/applier.go

```diff
@@ -0,0 +1,151 @@
+	receipt, penalty, reward, err := a.applyMessage(epoch, lm)
+
+	return vtypes.ApplyMessageResult{
+		Receipt: receipt,
+		Penalty: penalty,
+		Reward:  reward,
+		Root:    a.stateWrapper.Root().String(),
+	}, err
+}
+
+func (a *Applier) ApplySignedMessage(epoch abi.ChainEpoch, msg *vtypes.SignedMessage) (vtypes.ApplyMessageResult, error) {
+	var lm types.ChainMsg
+	switch msg.Signature.Type {
+	case acrypto.SigTypeSecp256k1:
+		lm = toLotusSignedMsg(msg)
+	case acrypto.SigTypeBLS:
+		lm = toLotusMsg(&msg.Message)
+	default:
+		return vtypes.ApplyMessageResult{}, errors.New("Unknown signature type")
+	}
+	// TODO: Validate the sig first
+	receipt, penalty, reward, err := a.applyMessage(epoch, lm)
+	return vtypes.ApplyMessageResult{
+		Receipt: receipt,
+		Penalty: penalty,
+		Reward:  reward,
+		Root:    a.stateWrapper.Root().String(),
+	}, err
+
+}
+
+func (a *Applier) ApplyTipSetMessages(epoch abi.ChainEpoch, blocks []vtypes.BlockMessagesInfo, rnd RandomnessSource) (vtypes.ApplyTipSetResult, error) {
+	cs := store.NewChainStore(a.stateWrapper.bs, a.stateWrapper.ds, a.syscalls)
+	sm := stmgr.NewStateManager(cs)
+
+	var bms []stmgr.BlockMessages
+	for _, b := range blocks {
+		bm := stmgr.BlockMessages{
+			Miner:    b.Miner,
+			WinCount: 1,
+		}
+
+		for _, m := range b.BLSMessages {
+			bm.BlsMessages = append(bm.BlsMessages, toLotusMsg(m))
+		}
+
+		for _, m := range b.SECPMessages {
+			bm.SecpkMessages = append(bm.SecpkMessages, toLotusSignedMsg(m))
+		}
+
+		bms = append(bms, bm)
+	}
+
+	var receipts []vtypes.MessageReceipt
+	sroot, _, err := sm.ApplyBlocks(context.TODO(), epoch-1, a.stateWrapper.Root(), bms, epoch, &randWrapper{rnd}, func(c cid.Cid, msg *types.Message, ret *vm.ApplyRet) error {
+		if msg.From == builtin.SystemActorAddr {
+			return nil // ignore reward and cron calls
+		}
+		rval := ret.Return
+		if rval == nil {
+			rval = []byte{} // chain validation tests expect empty arrays to not be nil...
+		}
+		receipts = append(receipts, vtypes.MessageReceipt{
+			ExitCode:    ret.ExitCode,
+			ReturnValue: rval,
+
+			GasUsed: vtypes.GasUnits(ret.GasUsed),
+		})
+		return nil
+	})
+	if err != nil {
+		return vtypes.ApplyTipSetResult{}, err
+	}
+
+	a.stateWrapper.stateRoot = sroot
+
+	return vtypes.ApplyTipSetResult{
+		Receipts: receipts,
+		Root:     a.stateWrapper.Root().String(),
+	}, nil
+}
+
+func (a *Applier) applyMessage(epoch abi.ChainEpoch, lm types.ChainMsg) (vtypes.MessageReceipt, abi.TokenAmount, abi.TokenAmount, error) {
+	ctx := context.TODO()
+	base := a.stateWrapper.Root()
+
+	lotusVM, err := vm.NewVM(base, epoch, &vmRand{}, a.stateWrapper.bs, a.syscalls, nil)
+	// need to modify the VM invoker to add the puppet actor
+	chainValInvoker := vm.NewInvoker()
+	chainValInvoker.Register(puppet.PuppetActorCodeID, puppet.Actor{}, puppet.State{})
+	lotusVM.SetInvoker(chainValInvoker)
+	if err != nil {
+		return vtypes.MessageReceipt{}, big.Zero(), big.Zero(), err
+	}
+
+	ret, err := lotusVM.ApplyMessage(ctx, lm)
```

Comment:
we may want to eventually keep the VM open across multiple actions rather than flushing between each one.

more generally, we probably want to factor out the use of lotusVM to be able not directly run our linked version of the VM in-process?

Reply:
I see your point about flushing on every action (i.e. on every message) - this is probably redundant... although I'd consider this an optimisation - the result should be the same, no matter if we keep one VM, or we create a new one based on the same state. Maybe the VM could be part of the `stateWrapper`, but for now I didn't want to introduce more than the bare minimum of changes.

I am not sure I follow what you mean with `to run our linked version of the VM` though?

Reply:
this code dep's to a specific version of lotus & spec actors, and we're going to be tied to that as our reference as that's the code we'll be directly executing when we call vm here. Probably an pre-optimization worry that isn't a big deal yet.

Reply:
I think it is fine to use the Lotus VM for now. I don't think we have in scope to implement a reference VM at this stage, so no way around not including specs-actors and Lotus VM and living with them changing under our feet (which at this stage really shouldn't happen often, or should be minimal).

I suggest we add an abstraction on the VM and inject it as soon as we have another impl. but not before.